src/web_clipboard.rs

```diff
@@ -0,0 +1,146 @@
+}
+
+/// To get data from web events
+#[derive(Default)]
+pub struct WebChannel<T> {
+    rx: Option<Receiver<T>>,
+}
+
+impl<T> WebChannel<T> {
+    /// Only returns Some if user explicitly triggered an event. Should be called each frame to react as soon as the event is fired.
+    pub fn try_read_clipboard_event(&mut self) -> Option<T> {
+        match &mut self.rx {
+            Some(rx) => {
+                if let Ok(data) = rx.try_recv() {
+                    return Some(data);
+                }
+                None
+            }
+            None => None,
+        }
+    }
+}
+
+/// User provided a string to paste
+#[derive(Debug, Default)]
+pub struct WebEventPaste(pub String);
+/// User asked to cut
+#[derive(Default)]
+pub struct WebEventCut;
+/// Used asked to copy
+#[derive(Default)]
+pub struct WebEventCopy;
+
+fn setup_clipboard_copy(clipboard_channel: &mut WebChannel<WebEventCopy>) {
+    let (tx, rx): (Sender<WebEventCopy>, Receiver<WebEventCopy>) = crossbeam_channel::bounded(1);
+
+    let closure = Closure::<dyn FnMut(_)>::new(move |event: web_sys::ClipboardEvent| {
+        // TODO: maybe we should check if current canvas is selected ? not sure it's possible,
+        // but reacting to event at the document level will lead to problems if multiple games are on the same page.
+        tx.try_send(WebEventCopy);
+    });
+
+    // TODO: a lot of unwraps ; it's using documents because paste event set on a canvas do not trigger (also tested on firefox in vanilla javascript)
+    let listener = closure.as_ref().unchecked_ref();
+    web_sys::window()
+        .unwrap()
+        .document()
+        .unwrap()
+        .add_event_listener_with_callback("copy", listener)
+        .expect("Could not add copy event listener.");
+    closure.forget();
+    *clipboard_channel = WebChannel::<WebEventCopy> { rx: Some(rx) };
+}
+
+fn setup_clipboard_cut(clipboard_channel: &mut WebChannel<WebEventCut>) {
+    let (tx, rx): (Sender<WebEventCut>, Receiver<WebEventCut>) = crossbeam_channel::bounded(1);
+
+    let closure = Closure::<dyn FnMut(_)>::new(move |event: web_sys::ClipboardEvent| {
+        // TODO: maybe we should check if current canvas is selected ? not sure it's possible,
+        // but reacting to event at the document level will lead to problems if multiple games are on the same page.
+        tx.try_send(WebEventCut);
+    });
+
+    // TODO: a lot of unwraps ; it's using documents because paste event set on a canvas do not trigger (also tested on firefox in vanilla javascript)
+    let listener = closure.as_ref().unchecked_ref();
+    web_sys::window()
+        .unwrap()
+        .document()
+        .unwrap()
+        .add_event_listener_with_callback("cut", listener)
+        .expect("Could not add cut event listener.");
+    closure.forget();
+    *clipboard_channel = WebChannel::<WebEventCut> { rx: Some(rx) };
+}
+
+fn setup_clipboard_paste(clipboard_channel: &mut WebChannel<WebEventPaste>) {
```

Comment:
all those functions are mostly similar, that doesn't feel so good duplicating them, but there's only 3 and I feel any genericity would complicate their signature quite a bit

Reply:
yeah, fine by me 👍